src/topology/config/validation.rs

```diff
@@ -185,17 +209,40 @@ impl From<&Config> for Graph {
     }
 }
 
-fn paths_rec(nodes: &HashMap<String, Node>, node: &str, mut path: Vec<String>) -> Vec<Vec<String>> {
+fn paths_rec(
+    nodes: &HashMap<String, Node>,
+    node: &str,
+    mut path: Vec<String>,
+) -> Result<Vec<Vec<String>>, String> {
+    if let Some(i) = path.iter().position(|p| p == node) {
+        let mut segment = path.split_off(i);
+        segment.push(node.into());
+        // I think this is maybe easier to grok from source -> sink, but I'm not
+        // married to either.
+        segment.reverse();
+        return Err(format!(
+            "Cyclic dependency detected in the chain [ {} ]",
+            segment.join(" -> ")
+        ));
+    }
     path.push(node.to_string());
     match nodes.get(node) {
         Some(Node::Source { .. }) | None => {
             path.reverse();
-            vec![path]
+            Ok(vec![path])
+        }
+        Some(Node::Transform { inputs, .. }) | Some(Node::Sink { inputs, .. }) => {
+            let mut paths = Vec::new();
+            for input in inputs {
+                match paths_rec(nodes, input, path.clone()) {
+                    Ok(mut p) => paths.append(&mut p),
+                    Err(err) => {
+                        return Err(err);
+                    }
+                }
+            }
```

Comment:
It _feels_ like there should be a way to do this with iterators still, but I haven't quite figured out what it is.